clash-vexriscv/src/ffi/impl.cpp

```diff
@@ -229,78 +230,110 @@ vexr_jtag_bridge_data *vexr_jtag_bridge_init(uint16_t port)
-				}
-			}
-		}
-	}
-	d->timer = 3;
+void vexr_jtag_bridge_step(vexr_jtag_bridge_data *d, const JTAG_OUTPUT *output, JTAG_INPUT *input) {
+    *input = d->prev_input;
+
+    if (d->timer != 0) {
+        d->timer -= 1;
+        return;
+    }
+
+    d->check_new_connections_timer++;
+    if (d->check_new_connections_timer == 200) {
+        d->check_new_connections_timer = 0;
+        int new_client_handle = accept(
+            d->server_socket,
+            (struct sockaddr *) &d->server_storage,
+            &d->addr_size
+        );
+        if (new_client_handle != -1) {
+            if (d->client_handle != -1) {
+                connection_reset(d);
+            }
+            d->client_handle = new_client_handle;
+        } else if (d->client_handle == -1) {
+            d->self_sleep = 200;
+        }
+    }
+
+    if (d->self_sleep) {
+        d->self_sleep--;
+        return;
+    }
+
+    if (d->client_handle != -1) {
+        int n;
+
+        if (d->rx_buffer_remaining == 0) {
+            if (ioctl(d->client_handle, FIONREAD, &n) != 0) {
+                connection_reset(d);
+                return;
+            }
+            if (n >= 1) {
+                d->rx_buffer_size = read(d->client_handle, &d->rx_buffer, sizeof(d->rx_buffer));
+                if (d->rx_buffer_size < 0) {
+                    connection_reset(d);
+                    return;
+                }
+                d->rx_buffer_remaining = d->rx_buffer_size;
+            } else {
+                d->self_sleep = 30;
+                return;
+            }
+        }
+
+        if (d->rx_buffer_remaining != 0) {
+            char command = d->rx_buffer[d->rx_buffer_size - (d->rx_buffer_remaining--)];
+            switch (command) {
+                case 'R': { // Read request
+                    char tdo_value = (output->tdo != 0) ? '1' : '0';
+                    if (send(d->client_handle, &tdo_value, 1, 0) == -1) {
+                        printf("[REMOTE_BITBANG] send failed\n");
+                        connection_reset(d);
+                    }
+
+                    break;
+                }
+                case '0' ... '7': { // Write tck, tms, tdi
+                    uint8_t value = command - '0';
+                    input->tck = (value & 4) != 0;
+                    input->tms = (value & 2) != 0;
+                    input->tdi = (value & 1) != 0;
+                    d->prev_input = *input;
+                    break;
+                }
+                case 'Q': { // Quit
+                    connection_reset(d);
+                    break;
+                }
+								case 'B': { // Blink on
+									break;
+								}
+								case 'b': { // Blink off
+									break;
+								}
+								case 'r': { // Reset 0 0
+									break;
+								}
+								case 's': { // Reset 0 1
+									break;
+								}
+								case 't': { // Reset 1 1
+									break;
+								}
+								case 'u': { // Reset 1 0
+									break;
+								}
```

Comment:
The heck happened here :see_no_evil:

Reply:
These are [remote-bitbang](https://github.com/openocd-org/openocd/blob/master/doc/manual/jtag/drivers/remote_bitbang.txt#L51-L73) commands. There's not really a way for us to blink other than maybe using `printf` .For the resets, I have not seen them being driven in simulation and don't know how we'd have to wire them up in simulation.

Reply:
I meant the weird indentation. It suddenly jumps by 40 characters or so.